File: src/main/java/com/thebluehats/server/game/enchants/Perun.java
package com.thebluehats.server.game.enchants;

import java.util.ArrayList;

import com.google.inject.Inject;
import com.thebluehats.server.game.managers.combat.DamageManager;
import com.thebluehats.server.game.managers.combat.templates.DamageEnchantTrigger;
import com.thebluehats.server.game.managers.combat.templates.EnchantHolder;
import com.thebluehats.server.game.managers.combat.templates.PlayerDamageTrigger;
import com.thebluehats.server.game.managers.enchants.DamageTriggeredEnchant;
import com.thebluehats.server.game.managers.enchants.EnchantGroup;
import com.thebluehats.server.game.managers.enchants.EnchantProperty;
import com.thebluehats.server.game.managers.enchants.HitCounter;
import com.thebluehats.server.game.managers.enchants.processedevents.DamageEventEnchantData;
import com.thebluehats.server.game.utils.EnchantLoreParser;
import com.thebluehats.server.game.utils.EntityValidator;

import org.bukkit.ChatColor;
import org.bukkit.Material;
import org.bukkit.entity.Player;

public class Perun extends DamageTriggeredEnchant {
    private final EnchantProperty<Integer> perunDamage = new EnchantProperty<>(3, 4, 2);
    private final EnchantProperty<Integer> hitsNeeded = new EnchantProperty<>(5, 4, 4);

    private final DamageManager damageManager;
    private final HitCounter hitCounter;

    @Inject
    public Perun(DamageManager damageManager, HitCounter hitCounter, PlayerDamageTrigger playerDamageTrigger) {
        super(new DamageEnchantTrigger[] { playerDamageTrigger }, new EntityValidator[] { damageManager });

        this.damageManager = damageManager;
        this.hitCounter = hitCounter;
    }

    @Override
    public void execute(DamageEventEnchantData data) {
        Player damager = data.getDamager();
        Player damagee = data.getDamagee();
        int level = data.getLevel();

        int damage = perunDamage.getValueAtLevel(level);

        hitCounter.addOne(damager);

        if (hitCounter.hasHits(damager, hitsNeeded.getValueAtLevel(level))) {
            if (level == 3) {
                if (damagee.getInventory().getBoots() != null)
                    if (damagee.getInventory().getBoots().getType() == Material.DIAMOND_BOOTS)
                        damage += 2;
                if (damagee.getInventory().getChestplate() != null)
                    if (damagee.getInventory().getChestplate().getType() == Material.DIAMOND_CHESTPLATE)
                        damage += 2;
                if (damagee.getInventory().getLeggings() != null)
                    if (damagee.getInventory().getLeggings().getType() == Material.DIAMOND_LEGGINGS)
                        damage += 2;
                if (damagee.getInventory().getHelmet() != null)
                    if (damagee.getInventory().getHelmet().getType() == Material.DIAMOND_HELMET)
                        damage += 2;
            }

            damager.getWorld().strikeLightningEffect(damagee.getLocation());
            damageManager.doTrueDamage(damagee, damage);
        }
    }

    @Override
    public String getName() {
        return "Combo: Perun's Wrath";
    }

    @Override
    public String getEnchantReferenceName() {
        return "Perun";
    }

    @Override
    public ArrayList<String> getDescription(int level) {
        EnchantLoreParser enchantLoreParser = new EnchantLoreParser("");

        String lastMessage = ChatColor.ITALIC + "Lightning deals true damage";

        enchantLoreParser.addTextIf(level != 3,
                "Every <yellow>fourth</yellow> hit strikes<br/>lightning for <red>{0}❤</red><br/>" + lastMessage);
        enchantLoreParser.addTextIf(level == 3,
                "Every <yellow>fourth</yellow> hit strikes<br/>lightning for <red>{0}❤</red> + <red>1❤</red><br/>per <aqua>diamond piece</aqua> on your<br/> victim<br/>"
                        + lastMessage);

        enchantLoreParser.setSingleVariable("1.5", "2", "1");

        return enchantLoreParser.parseForLevel(level);
    }

    @Override
    public boolean isDisabledOnPassiveWorld() {
        return false;
    }

    @Override
    public EnchantGroup getEnchantGroup() {
        return EnchantGroup.B;
    }

    @Override
    public boolean isRareEnchant() {
        return true;
    }

    @Override
    public Material[] getEnchantItemTypes() {
        return new Material[] { Material.GOLD_SWORD };
    }

    @Override
    public EnchantHolder getEnchantHolder() {
        return EnchantHolder.DAMAGER;
    }
}

File: src/main/java/com/thebluehats/server/game/enchants/Perun.kt
package com.thebluehats.server.game.enchants

import com.google.inject.Inject
import com.thebluehats.server.game.managers.combat.DamageManager
import com.thebluehats.server.game.managers.combat.templates.DamageEnchantTrigger
import com.thebluehats.server.game.managers.combat.templates.EnchantHolder
import com.thebluehats.server.game.managers.combat.templates.PlayerDamageTrigger
import com.thebluehats.server.game.managers.enchants.DamageTriggeredEnchant
import com.thebluehats.server.game.managers.enchants.EnchantGroup
import com.thebluehats.server.game.managers.enchants.EnchantProperty
import com.thebluehats.server.game.managers.enchants.HitCounter
import com.thebluehats.server.game.managers.enchants.processedevents.DamageEventEnchantData
import com.thebluehats.server.game.utils.EnchantLoreParser
import com.thebluehats.server.game.utils.EntityValidator
import org.bukkit.ChatColor
import org.bukkit.Material
import java.util.*

class Perun @Inject constructor(
    private val damageManager: DamageManager,
    private val hitCounter: HitCounter,
    playerDamageTrigger: PlayerDamageTrigger
) : DamageTriggeredEnchant(
    arrayOf(playerDamageTrigger), arrayOf(
        damageManager
    )
) {
    private val perunDamage = EnchantProperty(3, 4, 2)
    private val hitsNeeded = EnchantProperty(5, 4, 4)

    override val name: String get() = "Perun"
    override val enchantReferenceName: String get() = "Perun"
    override val isDisabledOnPassiveWorld: Boolean get() = false
    override val enchantGroup: EnchantGroup get() = EnchantGroup.B
    override val isRareEnchant: Boolean get() = true
    override val enchantItemTypes: Array<Material> get() = arrayOf(Material.GOLD_SWORD)
    override val enchantHolder: EnchantHolder get() = EnchantHolder.DAMAGER

    override fun execute(data: DamageEventEnchantData) {
        val damager = data.damager
        val damagee = data.damagee
        val level = data.level
        var damage = perunDamage.getValueAtLevel(level)
        hitCounter.addOne(damager)
        if (hitCounter.hasHits(damager, hitsNeeded.getValueAtLevel(level))) {
            if (level == 3) {
                if (damagee.inventory.boots != null) if (damagee.inventory.boots.type == Material.DIAMOND_BOOTS) damage += 2
                if (damagee.inventory.chestplate != null) if (damagee.inventory.chestplate.type == Material.DIAMOND_CHESTPLATE) damage += 2
                if (damagee.inventory.leggings != null) if (damagee.inventory.leggings.type == Material.DIAMOND_LEGGINGS) damage += 2
                if (damagee.inventory.helmet != null) if (damagee.inventory.helmet.type == Material.DIAMOND_HELMET) damage += 2
            }
            damager.world.strikeLightningEffect(damagee.location)
            damageManager.doTrueDamage(damagee, damage.toDouble(), damager)
        }
    }

    override fun getDescription(level: Int): ArrayList<String> {
        val enchantLoreParser = EnchantLoreParser("")
        val lastMessage = ChatColor.ITALIC.toString() + "Lightning deals true damage"
        enchantLoreParser.addTextIf(
            level == 2,
            "Every <yellow>fifth</yellow> hit strikes<br/><yellow>lightning</yellow> for <red>{0}❤</red><br/>$lastMessage"
        )
        enchantLoreParser.addTextIf(
            level == 2,
            "Every <yellow>fourth</yellow> hit strikes<br/><yellow>lightning</yellow> for <red>{0}❤</red><br/>$lastMessage"
        )
        enchantLoreParser.addTextIf(
            level == 3,
            "Every <yellow>fourth</yellow> hit strikes<br/><yellow>lightning</yellow> for <red>{0}❤</red> + " +
                "<red>1❤</red><br/>per <aqua>diamond piece</aqua> on your<br/>victim.<br/>"
                    + lastMessage
        )
        enchantLoreParser.setSingleVariable("1.5", "2", "1")
        return enchantLoreParser.parseForLevel(level)
    }
}